Mathlib/Order/Interval/Set/Icc.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2025 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll, Kevin Buzzard, Bhavik Mehta, Oliver Nash
+-/
+import Mathlib.Data.Fintype.Order
+import Mathlib.Data.Nat.Lattice
+import Mathlib.Data.Int.ConditionallyCompleteOrder
+import Mathlib.Data.Int.Interval
+import Mathlib.Data.Int.SuccPred
+
+/-!
+# Characterisation of the closed interval
+
+In this file we provide a characterisation of the closed interval for locally-finite conditionally
+complete linear orders, together with convenience specialisations for use with types such as
+`ℕ`, `ℤ`, `Fin n`.
+
+## Main results:
+ * `Set.Nonempty.eq_Icc_of_Ioo_subseteq_compl_empty`: a characterisation of closed intervals for
+   locally-finite conditionally complete linear orders.
+ * `Set.Nonempty.eq_Icc_of_Ioo_subseteq_compl_empty'`: characterisation of closed intervals for
+   locally-finite conditionally complete linear orders (convenient for `Fin n`).
+ * `Set.Finite.eq_Icc_of_Ioo_subseteq_compl_empty_nat`: characterisation of closed intervals
+   for `ℕ`.
+ * `Set.Finite.eq_Icc_of_Ioo_subseteq_compl_empty_int`: characterisation of closed intervals
+   for `ℤ`.
+-/
+
+variable {α : Type*} {I : Set α}
+
+lemma Set.Finite.ordConnected_of_Ioo_subseteq_compl_empty [LinearOrder α]
+    (h₁ : I.Finite)
+    (h₂ : ∀ᵉ (x ∈ I) (y ∈ I), Ioo x y ⊆ Iᶜ → Ioo x y = ∅) :
+    I.OrdConnected := by
```

Comment:
Is this theorem true as an `Iff`?

Reply:
If it is, then my preference is that the lemmas below be stated in terms of OrdConnected instead, as the "canonical" way of describing this condition

Reply:
I think the following might be the "right" lemma:
```lean
lemma Set.ordConnected_iff_disjoint_Ioo_empty [LinearOrder α] [LocallyFiniteOrder α] :
    I.OrdConnected ↔ ∀ᵉ (x ∈ I) (y ∈ I), Disjoint (Ioo x y) I → Ioo x y = ∅ := by
```
I expect to have some time this evening so I'll attempt to prove + use this, and adjust as you suggest.

Reply:
I'm happy to set this up however you think best: all I really want is the characterisation of closed intervals in `ℤ` which is convenient for my purpose.

For now I've pushed up a version assuming `LocallyFiniteOrder` but I do see your point. I think there is something more to this since I couldn't see a proof that `∀ᵉ (x ∈ I) (y ∈ I), Disjoint (Ioo x y) I → Ioo x y = ∅` and `I.OrdConnected` are equivalent without adding some assumptions, and `LocallyFiniteOrder` seems to be too strong (e.g., it rules out `ℝ`). However I'm inclined to think that this is enough general nonsense for the time being.

Reply:
Makes sense, would you mind leaving a comment/TODO about generalising the typeclasses here so it doesn't get forgotten? I'm otherwise happy to stick with your current version

Reply:
I've added a TODO try and capture this. LMK if you feel it needs adjusting.